File: src/test/java/de/spries/fleetcommander/model/facade/PlayerSpecificPlanetTest.java
package de.spries.fleetcommander.model.facade;

import de.spries.fleetcommander.model.core.Player;
import de.spries.fleetcommander.model.core.universe.Planet;
import de.spries.fleetcommander.model.core.universe.PlanetClass;
import org.junit.Before;
import org.junit.Test;

import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.nullValue;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class PlayerSpecificPlanetTest {

	private Planet originalPlanet;
	private Player self;
	private Player otherPlayer;
	private PlayerSpecificPlanet ownPlanet;
	private PlayerSpecificPlanet otherPlayersPlanet;

	@Before
	public void setUp() {
		originalPlanet = mock(Planet.class);
		self = mock(Player.class);
		otherPlayer = mock(Player.class);

		ownPlanet = new PlayerSpecificPlanet(originalPlanet, self);
		otherPlayersPlanet = new PlayerSpecificPlanet(originalPlanet, otherPlayer);

		doReturn(true).when(originalPlanet).isInhabitedBy(self);
		doReturn(false).when(originalPlanet).isInhabitedBy(otherPlayer);
	}

	@Test
	public void forwardsCallToGetId() {
		otherPlayersPlanet.getId();
		verify(originalPlanet).getId();
	}

	@Test
	public void forwardsCallToGetX() {
		otherPlayersPlanet.getX();
		verify(originalPlanet).getX();
	}

	@Test
	public void forwardsCallToGetY() {
		otherPlayersPlanet.getY();
		verify(originalPlanet).getY();
	}

	@Test
	public void forwardsCallToPlanetClassForSelf() {
		when(originalPlanet.getPlanetClass()).thenReturn(PlanetClass.P);
		ownPlanet.getPlanetClass();
		verify(originalPlanet).getPlanetClass();
	}

	@Test
	public void doesNotReturnPlanetClassForOtherPlayers() {
		assertThat(otherPlayersPlanet.getPlanetClass(), is("?"));
		verify(originalPlanet, never()).getPlanetClass();
	}

	@Test
	public void forwardsCallToIsHomePlanet() {
		ownPlanet.isMyHomePlanet();
		verify(originalPlanet).isHomePlanetOf(self);
	}

	@Test
	public void forwardsCallToIsInhabitedBy() {
		ownPlanet.isInhabitedByMe();
		verify(originalPlanet).isInhabitedBy(self);
	}

	@Test
	public void forwardsCallToIsKnownAsEnemyPlanet() {
		ownPlanet.isKnownAsEnemyPlanet();
		verify(originalPlanet).isKnownAsEnemyPlanet(self);
	}

	@Test
	public void forwardsCallToIsUnderAttackForSelf() {
		ownPlanet.isUnderAttack();
		verify(originalPlanet).isUnderAttack();
	}

	@Test
	public void doesNotReturnIsUnderAttackForOtherPlayers() {
		assertThat(otherPlayersPlanet.isUnderAttack(), is(false));
		verify(originalPlanet, never()).isUnderAttack();
	}

	@Test
	public void forwardsCallToIsJustInhabitedForSelf() {
		ownPlanet.isJustInhabited();
		verify(originalPlanet).isJustInhabited();
	}

	@Test
	public void doesNotReturnIsJustInhabitedForOtherPlayers() {
		assertThat(otherPlayersPlanet.isJustInhabited(), is(false));
		verify(originalPlanet, never()).isJustInhabited();
	}

	@Test
	public void forwardsCallToCanBuildFactory() throws Exception {
		ownPlanet.canBuildFactory();
		verify(originalPlanet).canBuildFactory(self);
	}

	@Test
	public void forwardsCallToGetShipCountForSelf() {
		doReturn(true).when(originalPlanet).isInhabitedBy(self);

		ownPlanet.getShipCount();
		verify(originalPlanet).getShipCount();
	}

	@Test
	public void doesNotReturnShipCountForOtherPlayers() {
		assertThat(otherPlayersPlanet.getShipCount(), is(0));
		verify(originalPlanet, never()).getShipCount();
	}

	@Test
	public void forwardsCallToGetIncomingShips() {
		ownPlanet.getIncomingShipCount();
		verify(originalPlanet).getIncomingShipCount(self);
	}

	@Test
	public void forwardsCallToGetFactorySiteForSelf() {
		ownPlanet.getFactorySite();
		verify(originalPlanet).getFactorySite();
	}

	@Test
	public void doesNotReturnFactorySiteForOtherPlayers() {
		assertThat(otherPlayersPlanet.getFactorySite(), is(nullValue()));
		verify(originalPlanet, never()).getFactorySite();
	}

	@Test
	public void forwardsCallToBuildFactory() {
		ownPlanet.buildFactory();
		verify(originalPlanet).buildFactory(self);
	}

	@Test
	public void forwardsCallToSetProductionFocus() {
		ownPlanet.changeProductionFocus(1);
		verify(originalPlanet).setProductionFocus(1, self);
	}

}

File: src/test/kotlin/de/spries/fleetcommander/model/facade/PlayerSpecificPlanetTest.kt
package de.spries.fleetcommander.model.facade

import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import de.spries.fleetcommander.model.core.Player
import de.spries.fleetcommander.model.core.universe.Planet
import de.spries.fleetcommander.model.core.universe.PlanetClass
import org.hamcrest.Matchers.`is`
import org.hamcrest.Matchers.nullValue
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito.`when`
import org.mockito.Mockito.never
import org.mockito.Mockito.verify

class PlayerSpecificPlanetTest {

    private lateinit var originalPlanet: Planet
    private lateinit var self: Player
    private lateinit var otherPlayer: Player
    private lateinit var ownPlanet: PlayerSpecificPlanet
    private lateinit var otherPlayersPlanet: PlayerSpecificPlanet

    @Before
    fun setUp() {
        originalPlanet = mock()
        self = mock()
        otherPlayer = mock()

        ownPlanet = PlayerSpecificPlanet(originalPlanet, self)
        otherPlayersPlanet = PlayerSpecificPlanet(originalPlanet, otherPlayer)

        whenever(originalPlanet.isInhabitedBy(self)).thenReturn(true)
        whenever(originalPlanet.isInhabitedBy(otherPlayer)).thenReturn(false)
    }

    @Test
    fun forwardsCallToGetId() {
        otherPlayersPlanet.getId()
        verify(originalPlanet).id
    }

    @Test
    fun forwardsCallToPlanetClassForSelf() {
        `when`(originalPlanet.getPlanetClass()).thenReturn(PlanetClass.P)
        ownPlanet.getPlanetClass()
        verify(originalPlanet).getPlanetClass()
    }

    @Test
    fun doesNotReturnPlanetClassForOtherPlayers() {
        assertThat(otherPlayersPlanet.getPlanetClass(), `is`("?"))
        verify(originalPlanet, never()).getPlanetClass()
    }

    @Test
    fun forwardsCallToIsHomePlanet() {
        ownPlanet.isMyHomePlanet()
        verify(originalPlanet).isHomePlanetOf(self)
    }

    @Test
    fun isHomePlanetWhenOwnHomePlanet() {
        `when`(originalPlanet.isHomePlanetOf(self)).thenReturn(true)
        assertThat(ownPlanet.isHomePlanet(), `is`(true))
    }

    @Test
    fun isNoHomePlanetWhenNotHomePlanetAtAll() {
        `when`(originalPlanet.isHomePlanetOf(self)).thenReturn(false)
        `when`(originalPlanet.isHomePlanet()).thenReturn(false)
        assertThat(ownPlanet.isHomePlanet(), `is`(false))
    }

    @Test
    fun isHomePlanetWhenOtherPlayersHomePlanetAndVisited() {
        `when`(originalPlanet.isHomePlanetOf(self)).thenReturn(false)
        `when`(originalPlanet.isHomePlanet()).thenReturn(true)
        `when`(originalPlanet.isKnownAsEnemyPlanet(self)).thenReturn(true)
        assertThat(ownPlanet.isHomePlanet(), `is`(true))
    }

    @Test
    fun isNoHomePlanetWhenOtherPlayersHomePlanetAndNotVisited() {
        `when`(originalPlanet.isHomePlanetOf(self)).thenReturn(false)
        `when`(originalPlanet.isHomePlanet()).thenReturn(true)
        `when`(originalPlanet.isKnownAsEnemyPlanet(self)).thenReturn(false)
        assertThat(ownPlanet.isHomePlanet(), `is`(false))
    }

    @Test
    fun forwardsCallToIsInhabitedBy() {
        ownPlanet.isInhabitedByMe()
        verify(originalPlanet).isInhabitedBy(self)
    }

    @Test
    fun forwardsCallToIsKnownAsEnemyPlanet() {
        ownPlanet.isKnownAsEnemyPlanet()
        verify(originalPlanet).isKnownAsEnemyPlanet(self)
    }

    @Test
    fun forwardsCallToIsUnderAttackForSelf() {
        ownPlanet.isUnderAttack()
        verify(originalPlanet).isUnderAttack()
    }

    @Test
    fun doesNotReturnIsUnderAttackForOtherPlayers() {
        assertThat(otherPlayersPlanet.isUnderAttack(), `is`(false))
        verify(originalPlanet, never()).isUnderAttack()
    }

    @Test
    fun forwardsCallToIsJustInhabitedForSelf() {
        ownPlanet.isJustInhabited()
        verify(originalPlanet).isJustInhabited()
    }

    @Test
    fun doesNotReturnIsJustInhabitedForOtherPlayers() {
        assertThat(otherPlayersPlanet.isJustInhabited(), `is`(false))
        verify(originalPlanet, never()).isJustInhabited()
    }

    @Test
    fun forwardsCallToCanBuildFactory() {
        ownPlanet.canBuildFactory()
        verify(originalPlanet).canBuildFactory(self)
    }

    @Test
    fun forwardsCallToGetShipCountForSelf() {
        whenever(originalPlanet.isInhabitedBy(self)).thenReturn(true)

        ownPlanet.getShipCount()
        verify(originalPlanet).getShipCount()
    }

    @Test
    fun doesNotReturnShipCountForOtherPlayers() {
        assertThat(otherPlayersPlanet.getShipCount(), `is`(0))
        verify(originalPlanet, never()).getShipCount()
    }

    @Test
    fun forwardsCallToGetIncomingShips() {
        ownPlanet.getIncomingShipCount()
        verify(originalPlanet).getIncomingShipCount(self)
    }

    @Test
    fun forwardsCallToGetFactorySiteForSelf() {
        ownPlanet.getFactorySite()
        verify(originalPlanet).getFactorySite()
    }

    @Test
    fun doesNotReturnFactorySiteForOtherPlayers() {
        assertThat(otherPlayersPlanet.getFactorySite(), `is`(nullValue()))
        verify(originalPlanet, never()).getFactorySite()
    }

    @Test
    fun forwardsCallToBuildFactory() {
        ownPlanet.buildFactory()
        verify(originalPlanet).buildFactory(self)
    }

    @Test
    fun forwardsCallToSetProductionFocus() {
        ownPlanet.changeProductionFocus(1)
        verify(originalPlanet).setProductionFocus(1, self)
    }

}
